### PR TITLE
[Geo] Add Well Known Text (WKT) Parsing Support to ShapeBuilders

### LINKED ISSUE #9120: Support WKT syntax in geo_point and geo_shape
So indexing stuff like `POINT (30 10)` and `POLYGON ((30 10, 40 40, 20 40, 10 20, 30 10))` can be done into geo_point and geo_shape field types

### PR DESCRIPTION
**Pull request by nknize**
This PR adds WKT support to Geo ShapeBuilders.

This supports the following format:

```
POINT (30 10)
LINESTRING (30 10, 10 30, 40 40)
POLYGON ((30 10, 40 40, 20 40, 10 20, 30 10))
POLYGON ((35 10, 45 45, 15 40, 10 20, 35 10), (20 30, 35 35, 30 20, 20 30))
MULTIPOINT ((10 40), (40 30), (20 20), (30 10))
MULTIPOINT (10 40, 40 30, 20 20, 30 10)
MULTILINESTRING ((10 10, 20 20, 10 40),(40 40, 30 30, 40 20, 30 10))
MULTIPOLYGON (((30 20, 45 40, 10 40, 30 20)), ((15 5, 40 10, 10 20, 5 10, 15 5)))
MULTIPOLYGON (((40 40, 20 45, 45 30, 40 40)), ((20 35, 10 30, 10 10, 30 5, 45 20, 20 35), (30 20, 20 15, 20 25, 30 20)))
```
closes #9120

### CODE REVIEW

**Comment by nknize** [comment]
@elasticmachine please test this.

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: This will also accept something like "BBOX (-10) 10) -10) 10)", I think this should throw an error. Its a bit constructed but I think the parser shouldn't be that lenient.

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: Is this a potential Z value? So we support parsing it but drop it?

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: Close or Comma will also again allow for things like "( 1, 2, 3) 4 )" being accepted, no?

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: What needs to be done here?

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: What if next word is something else than LPAREN?

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: same problem as above I think, only that unmatched closing brackets will create holes here.

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: nit: whitespace

**Review by cbuescher — CHANGES_REQUESTED** [review]
@nknize thanks, I did a first round of review. Most changes in the builders look fine, however I think the WKT parser currently is a bit too lenient in what it accepts as valid WKT values (I left one example, but I think there are many more edge cases that are not currently tested and should be rejected).
It would also be great to test the parser not only with well formed values that the current builders output but add test for values that should be rejected for various reasons. Not sure how the current parser can be tweaked to do this, but I think it would be good to add this before this can be merged.

File: core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (nknize) [reply core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: This is used for testing `LineStringBuilder` and `MultiLineStringBuilder`. If I allow points to cross the dateline, a `MultiLineStringBuilder` is always created. To keep testing simple, I limited the random generation of linestrings to not cross the dateline. This gave me control over testing `LineStrings` and `MultiLineStrings` separately.

File: core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (nknize) [reply core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: Simple fix... the random point generation was set to `randomIntBetween(0, 100)`. 0 will throw an empty/invalid points exception.

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (nknize) [reply core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: I actually meant closer. As in a closer parentheses. But will change if this is confusing.

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (nknize) [reply core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: Actually, nothing. This is handled in the builder. Removing todo.

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (nknize) [reply core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: :+1:

File: core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilder.java
Comment (nknize) [reply core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilder.java]: Not yet... will be handled in #25738

File: core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilder.java
Comment (nknize) [reply core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilder.java]: :+1:

File: core/src/main/java/org/elasticsearch/common/geo/builders/GeometryCollectionBuilder.java
Comment (nknize) [reply core/src/main/java/org/elasticsearch/common/geo/builders/GeometryCollectionBuilder.java]: bonehead typo

File: core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (nknize) [reply core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: Fixed... the random point generation was set to randomIntBetween(0, 100). 0 will throw an empty/invalid points exception.

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (cbuescher) [reply core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: Ok, I see. Nevermind, since its a private method I leave it up to you to change, I was confused a bit but the method is short enough to understand what its doing.

File: core/src/main/java/org/elasticsearch/common/geo/GeoShapeType.java
Comment (nknize) [reply core/src/main/java/org/elasticsearch/common/geo/GeoShapeType.java]: :+1:

File: core/src/main/java/org/elasticsearch/common/geo/GeoShapeType.java
Comment (nknize) [reply core/src/main/java/org/elasticsearch/common/geo/GeoShapeType.java]: :+1:

File: core/src/main/java/org/elasticsearch/common/geo/GeoShapeType.java
Comment (nknize) [reply core/src/main/java/org/elasticsearch/common/geo/GeoShapeType.java]: :+1:

File: core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilder.java
Comment (nknize) [reply core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilder.java]: Per the WKT spec, casing doesn't matter

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (nknize) [reply core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: Cleaned up coordinate list parsing to better handle coordinates with and without parens.

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (nknize) [reply core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: :+1:  Just push a commit to test for comments

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (nknize) [reply core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: :+1:  done

**Comment by nknize** [comment]
Thanks for the review @cbuescher 

I went ahead and added more strict parsing (to include better paren parsing for coordinate lists). I also added tests for comments and illegal geometry types. I think this is close. If you wouldn't mind giving another review. I will try to add geometry collection testing before committing.

Also, since this is a fresh feature, what we can do is set this as experimental for 6.1 and flush it out in bug fixes. Thoughts?

File: core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: Should this also work with holes? I just tried parsing the second polygon example from the issue decription and it fails (`POLYGON ((35 10, 45 45, 15 40, 10 20, 35 10), (20 30, 35 35, 30 20, 20 30))`)  I suspect the second argument is a hole so I was trying to see of this was tested, seems like this needs to be added still.

File: core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: I would prefer at least one good and one malformed case per test run. Otherwise we only catch bigger errors on multiple repetions and not e.g. on PR tests.

**Review by cbuescher — COMMENTED** [review]
@nknize I took another look and left two comments. In general I would still like to have more (and maybe not randomized, but simple edge cases) malformed tests, maybe a few typical cases per shape. Also Polygons with holes don't seem to get parsed correctly and I think tests are also missing for those cases.

File: core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (nknize) [reply core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: :+1:  Good catch! I agree we should have tests for holes. This is a widely used use case. Will add that next.

File: core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (nknize) [reply core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: I'll add an explicit malformed test for common typos.

**Comment by nknize** [comment]
@cbuescher just added explicit tests for the following:
* polygon with hole (can add one for multi polygon w/ and w/o holes)
* a malformed poly caused by common typos (can expand this test to include other common typos if we'd like).
* self crossing polygon

Will update documentation but I think this is close to ready. Let me know what you think.

**Review by cbuescher — COMMENTED** [review]
@nknize thanks, I left two more comments regarding tests. Also, since you mentioned docs, I think at least a short info about the possibility to use WKT for shapes should be part of this PR as well.

File: core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (nknize) [reply core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: :+1:

File: core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (nknize) [reply core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: I added an explicit test for polygons with holes below in `testParsePolygonWithHole`, are you asking for a random polygon with random hole? That's a bit more tricky to generate a random polygon with a random inscribed polygon as a hole. Probably not really necessary as long as we've tested at least one polygon with a hole?

File: core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (cbuescher) [reply core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: thanks

File: core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (cbuescher) [reply core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: Yes, I saw the explicit test and I was hoping for a randomized polygon/hole thing, but I just realized that also GeoJsonShapeParserTests doesn't do this fully randomized. So if it is difficult I'm also fine if there is only the explicit test.

**Comment by cbuescher** [comment]
@nknize thanks for updating the test, if the polygon + hole randomization is tricky, let's leave that out. I think this is ready once some documentation is added.

File: core/src/main/java/org/elasticsearch/common/geo/GeoUtils.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/GeoUtils.java]: Why do we need the fieldname as part of the parsed bbox structure. As far as I can see this gets only parsed when an inner `field` parameter is seen, the use of which seems both undocumented and it also doesn't get rendered when using the GeoBoundingBoxQueryBuilder#doXContent method. I have a hope that we can remove this maybe?

File: core/src/main/java/org/elasticsearch/common/geo/GeoUtils.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/GeoUtils.java]: Should this inner `field` parameter really still supported? I cannot find it mentioned in the docs and it doesn't seem to get written when using one of our builders. As far as I understand the whole bbox json should look like this:
```
"some.field.foo" : {
                        ... various formats defining the box ...
}
```

This is probably legacy and should be removed (probably in another PR?)

**Review by cbuescher — CHANGES_REQUESTED** [review]

**Comment by cbuescher** [comment]
@nknize I just did another round of reviews on the last commit adding WKT parsing support to the GeoBoundingBoxQueryBuilder. Does this need to be part of this PR? I agree it should be added to complete the WKT support (don't know what needs to be done with the other query builders that can contain shapes), but since the existing PR is already big as is (and was close to merging when the documentation is added)  I'd probably do this in a separate PR.

**Comment by nknize** [comment]
@cbuescher I agree. After I made the push I figured it would be best to do it in a separate PR. So I'll go ahead and do that. Going to commit the updated docs here shortly and consider this PR done (after the doc review).

**Comment by nknize** [comment]
@cbuescher added some WKT documentation to the `geo-shape.asciidoc`. If you're good with this I'll commit to 6.2 and 7.0 and open a separate PR for adding WKT support to GeoBoundingBox queries. Then move back to Z value support 
/cc @clintongormley

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (cbuescher) [inline docs/reference/mapping/types/geo-shape.asciidoc]: `gradle :docs:check` fails here, maybe because of the line break. Not sure how to fix though.

**Review by cbuescher — COMMENTED** [review]
@nknize thanks for pulling out the query builder changes, left a short question for the latest docs commit and apparently some console example still needs fixing, but other than that it looks good to me. Will give the rest another quick look but I think once the docs test passes this is good to merge.

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (clintongormley) [reply docs/reference/mapping/types/geo-shape.asciidoc]: If the line break is the problem, it's ok to have long lines.  The code blocks on the docs will scroll

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: I get a warning here that GEOMETRYCOLLECTION is missing, looking at the added documentation this should be supported I think. If it is missing here as a type, we also need a test for it to catch this.

File: core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java]: nit: might throw in a DEFAULT case instead.

File: core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: I think we need that test.

File: core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: Can this be randomized so the GeometryCollectionBuilder has between 0 and e.g. 2 or 3 shapes?

**Review by cbuescher — COMMENTED** [review]
@nknize I took a brief look at the added geometry collection support and left a small suggstion on improving the test. Also documentation tests still need fixing, I think after that this PR is ready.

**Comment by nknize** [comment]
Thanks for such a great review @cbuescher  Looks like I've fixed the doc failures (at least on my test run, we'll see what elasticmachine has to say) and I added some randomization to GeometryCollection test. Pending everything passes green, I think this is good to go. Let me know what you think.

**Comment by nknize** [comment]
@elasticmachine please test this

File: core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: Great to have this, but it doesn't seem to test the no-shapes case. Would be nice to add this as another case (maybe wrapped in  `if (rarely())`, we don't need to test that edge case very often.

**Review by cbuescher — APPROVED** [review]
@nknize thanks for adding the randomization to the test, I left a small note regarding the test simplification. The rest looks good, and since the doc tests seem to pass I think this is ready to be merged once you get a clear CI pass.

**Comment by synhershko** [comment]
Yay! thanks!

**Pull request merged**
